From 0513227e180be8974e3a65a9dc6cdabdfa0b5a61 Mon Sep 17 00:00:00 2001
From: Sven van Haastregt <sven.vanhaastregt@arm.com>
Date: Wed, 15 Jun 2022 11:54:46 +0100
Subject: [PATCH 055/172] Add SPV_KHR_subgroup_rotate support

Add support for mapping the SPV_KHR_subgroup_rotate extension
operations to and from SPIR-V friendly IR, as well as to and from the
cl_khr_subgroup_rotate OpenCL extension.

Specifications:
 - https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/KHR/SPV_KHR_subgroup_rotate.html
 - https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#cl_khr_subgroup_rotate
(cherry picked from commit a49de2a30cb31cd4e2e7643eb62f5f98feae1756)
---
 include/LLVMSPIRVExtensions.inc             |   1 +
 lib/SPIRV/OCLUtil.cpp                       |   5 +
 lib/SPIRV/SPIRVToOCL.cpp                    |  16 ++
 lib/SPIRV/SPIRVToOCL.h                      |   3 +
 lib/SPIRV/SPIRVUtil.cpp                     |   4 +
 lib/SPIRV/libSPIRV/SPIRVInstruction.h       |  18 ++
 lib/SPIRV/libSPIRV/SPIRVOpCode.h            |   7 +-
 lib/SPIRV/libSPIRV/SPIRVOpCodeEnum.h        |   1 +
 test/transcoding/SPV_KHR_subgroup_rotate.cl | 275 ++++++++++++++++++++
 9 files changed, 328 insertions(+), 2 deletions(-)
 create mode 100644 test/transcoding/SPV_KHR_subgroup_rotate.cl

diff --git a/include/LLVMSPIRVExtensions.inc b/include/LLVMSPIRVExtensions.inc
index 9f4e2fc8..8eaf24de 100644
--- a/include/LLVMSPIRVExtensions.inc
+++ b/include/LLVMSPIRVExtensions.inc
@@ -11,6 +11,7 @@ EXT(SPV_KHR_expect_assume)
 EXT(SPV_KHR_integer_dot_product)
 EXT(SPV_KHR_bit_instructions)
 EXT(SPV_KHR_uniform_group_instructions)
+EXT(SPV_KHR_subgroup_rotate)
 EXT(SPV_INTEL_subgroups)
 EXT(SPV_INTEL_media_block_io)
 EXT(SPV_INTEL_device_side_avc_motion_estimation)
diff --git a/lib/SPIRV/OCLUtil.cpp b/lib/SPIRV/OCLUtil.cpp
index ac4f63c7..540c28ae 100644
--- a/lib/SPIRV/OCLUtil.cpp
+++ b/lib/SPIRV/OCLUtil.cpp
@@ -424,6 +424,9 @@ template <> void SPIRVMap<std::string, Op, SPIRVInstruction>::init() {
   // cl_khr_subgroup_shuffle_relative
   _SPIRV_OP(group_shuffle_up, GroupNonUniformShuffleUp)
   _SPIRV_OP(group_shuffle_down, GroupNonUniformShuffleDown)
+  // cl_khr_subgroup_rotate
+  _SPIRV_OP(group_rotate, GroupNonUniformRotateKHR)
+  _SPIRV_OP(group_clustered_rotate, GroupNonUniformRotateKHR)
   // cl_khr_extended_bit_ops
   _SPIRV_OP(bitfield_insert, BitFieldInsert)
   _SPIRV_OP(bitfield_extract_signed, BitFieldSExtract)
@@ -1291,6 +1294,8 @@ public:
         else if (NameRef.contains("bit_extract")) {
           addUnsignedArgs(0, 1);
         }
+      } else if (NameRef.startswith("sub_group_clustered_rotate")) {
+        addUnsignedArg(2);
       } else if (NameRef.contains("shuffle") || NameRef.contains("clustered"))
         addUnsignedArg(1);
     } else if (NameRef.startswith("bitfield_insert")) {
diff --git a/lib/SPIRV/SPIRVToOCL.cpp b/lib/SPIRV/SPIRVToOCL.cpp
index 33da13e9..33384d1f 100644
--- a/lib/SPIRV/SPIRVToOCL.cpp
+++ b/lib/SPIRV/SPIRVToOCL.cpp
@@ -481,7 +481,23 @@ std::string SPIRVToOCLBase::getBallotBuiltinName(CallInst *CI, Op OC) {
   return Prefix + kSPIRVName::GroupPrefix + "ballot_" + GroupOp;
 }
 
+std::string SPIRVToOCLBase::getRotateBuiltinName(CallInst *CI, Op OC) {
+  assert((OC == OpGroupNonUniformRotateKHR) &&
+         "Not intended to handle other opcodes");
+  std::string Prefix = getGroupBuiltinPrefix(CI);
+  assert((Prefix == kOCLBuiltinName::SubPrefix) &&
+         "Workgroup scope is not supported for OpGroupNonUniformRotateKHR");
+
+  std::string OptionalClustered;
+  if (CI->arg_size() == 4)
+    OptionalClustered = "clustered_";
+  return Prefix + kSPIRVName::GroupPrefix + OptionalClustered + "rotate";
+}
+
 std::string SPIRVToOCLBase::groupOCToOCLBuiltinName(CallInst *CI, Op OC) {
+  if (OC == OpGroupNonUniformRotateKHR)
+    return getRotateBuiltinName(CI, OC);
+
   auto FuncName = OCLSPIRVBuiltinMap::rmap(OC);
   assert(FuncName.find(kSPIRVName::GroupPrefix) == 0);
 
diff --git a/lib/SPIRV/SPIRVToOCL.h b/lib/SPIRV/SPIRVToOCL.h
index a074539f..9712d126 100644
--- a/lib/SPIRV/SPIRVToOCL.h
+++ b/lib/SPIRV/SPIRVToOCL.h
@@ -253,6 +253,9 @@ public:
   /// example: GroupNonUniformBallotBitCount(Reduce) =>
   /// group_ballot_bit_count_iadd => sub_group_ballot_bit_count
   std::string getBallotBuiltinName(CallInst *CI, Op OC);
+  /// Transform OpGroupNonUniformRotateKHR to corresponding OpenCL function
+  /// name.
+  std::string getRotateBuiltinName(CallInst *CI, Op OC);
   /// Transform group opcode to corresponding OpenCL function name
   std::string groupOCToOCLBuiltinName(CallInst *CI, Op OC);
   /// Transform SPV-IR image opaque type into OpenCL representation,
diff --git a/lib/SPIRV/SPIRVUtil.cpp b/lib/SPIRV/SPIRVUtil.cpp
index 47aa29fe..024072b2 100644
--- a/lib/SPIRV/SPIRVUtil.cpp
+++ b/lib/SPIRV/SPIRVUtil.cpp
@@ -2169,6 +2169,10 @@ public:
     case OpGroupNonUniformShuffleDown:
       addUnsignedArg(2);
       break;
+    case OpGroupNonUniformRotateKHR:
+      if (ArgTys.size() == 4)
+        addUnsignedArg(3);
+      break;
     case OpGroupNonUniformInverseBallot:
     case OpGroupNonUniformBallotFindLSB:
     case OpGroupNonUniformBallotFindMSB:
diff --git a/lib/SPIRV/libSPIRV/SPIRVInstruction.h b/lib/SPIRV/libSPIRV/SPIRVInstruction.h
index e4b51147..4d0c96fb 100644
--- a/lib/SPIRV/libSPIRV/SPIRVInstruction.h
+++ b/lib/SPIRV/libSPIRV/SPIRVInstruction.h
@@ -2533,6 +2533,24 @@ _SPIRV_OP(GroupNonUniformShuffleUp, true, 6)
 _SPIRV_OP(GroupNonUniformShuffleDown, true, 6)
 #undef _SPIRV_OP
 
+class SPIRVGroupNonUniformRotateKHRInst : public SPIRVInstTemplateBase {
+public:
+  SPIRVCapVec getRequiredCapability() const override {
+    return getVec(CapabilityGroupNonUniformRotateKHR);
+  }
+
+  llvm::Optional<ExtensionID> getRequiredExtension() const override {
+    return ExtensionID::SPV_KHR_subgroup_rotate;
+  }
+};
+
+#define _SPIRV_OP(x, ...)                                                      \
+  typedef SPIRVInstTemplate<SPIRVGroupNonUniformRotateKHRInst, Op##x,          \
+                            __VA_ARGS__>                                       \
+      SPIRV##x;
+_SPIRV_OP(GroupNonUniformRotateKHR, true, 6, true)
+#undef _SPIRV_OP
+
 class SPIRVBlockingPipesIntelInst : public SPIRVInstTemplateBase {
 protected:
   SPIRVCapVec getRequiredCapability() const override {
diff --git a/lib/SPIRV/libSPIRV/SPIRVOpCode.h b/lib/SPIRV/libSPIRV/SPIRVOpCode.h
index 5cc77a3c..35062463 100644
--- a/lib/SPIRV/libSPIRV/SPIRVOpCode.h
+++ b/lib/SPIRV/libSPIRV/SPIRVOpCode.h
@@ -141,7 +141,9 @@ inline bool isAccessChainOpCode(Op OpCode) {
 inline bool hasExecScope(Op OpCode) {
   unsigned OC = OpCode;
   return (OpGroupWaitEvents <= OC && OC <= OpGroupSMax) ||
-         (OpGroupReserveReadPipePackets <= OC && OC <= OpGroupCommitWritePipe);
+         (OpGroupReserveReadPipePackets <= OC &&
+          OC <= OpGroupCommitWritePipe) ||
+         (OC == OpGroupNonUniformRotateKHR);
 }
 
 inline bool hasGroupOperation(Op OpCode) {
@@ -179,7 +181,8 @@ inline bool isGroupOpCode(Op OpCode) {
 
 inline bool isGroupNonUniformOpcode(Op OpCode) {
   unsigned OC = OpCode;
-  return OpGroupNonUniformElect <= OC && OC <= OpGroupNonUniformQuadSwap;
+  return (OpGroupNonUniformElect <= OC && OC <= OpGroupNonUniformQuadSwap) ||
+         (OC == OpGroupNonUniformRotateKHR);
 }
 
 inline bool isMediaBlockINTELOpcode(Op OpCode) {
diff --git a/lib/SPIRV/libSPIRV/SPIRVOpCodeEnum.h b/lib/SPIRV/libSPIRV/SPIRVOpCodeEnum.h
index e8c3a4ad..9d7da279 100644
--- a/lib/SPIRV/libSPIRV/SPIRVOpCodeEnum.h
+++ b/lib/SPIRV/libSPIRV/SPIRVOpCodeEnum.h
@@ -328,6 +328,7 @@ _SPIRV_OP(GroupNonUniformBitwiseXor, 361)
 _SPIRV_OP(GroupNonUniformLogicalAnd, 362)
 _SPIRV_OP(GroupNonUniformLogicalOr, 363)
 _SPIRV_OP(GroupNonUniformLogicalXor, 364)
+_SPIRV_OP(GroupNonUniformRotateKHR, 4431)
 _SPIRV_OP(SDotKHR, 4450)
 _SPIRV_OP(UDotKHR, 4451)
 _SPIRV_OP(SUDotKHR, 4452)
diff --git a/test/transcoding/SPV_KHR_subgroup_rotate.cl b/test/transcoding/SPV_KHR_subgroup_rotate.cl
new file mode 100644
index 00000000..9cc3c709
--- /dev/null
+++ b/test/transcoding/SPV_KHR_subgroup_rotate.cl
@@ -0,0 +1,275 @@
+// RUN: %clang_cc1 -triple spir-unknown-unknown -O1 -cl-std=CL2.0 -fdeclare-opencl-builtins -finclude-default-header -emit-llvm-bc %s -o %t.bc
+// RUN: llvm-spirv --spirv-ext=+SPV_KHR_subgroup_rotate %t.bc -o %t.spv
+// RUN: llvm-spirv --spirv-ext=+SPV_KHR_subgroup_rotate %t.spv -to-text -o %t.spt
+// RUN: FileCheck < %t.spt %s --check-prefix=CHECK-SPIRV
+
+// RUN: llvm-spirv -r %t.spv -o %t.rev.bc
+// RUN: llvm-dis < %t.rev.bc | FileCheck %s --check-prefixes=CHECK-COMMON,CHECK-LLVM
+// RUN: llvm-spirv -r %t.spv --spirv-target-env=SPV-IR -o %t.rev.bc
+// RUN: llvm-dis < %t.rev.bc | FileCheck %s --check-prefixes=CHECK-COMMON,CHECK-SPV-IR
+
+// From SPIR-V friendly IR:
+// RUN: llvm-spirv %t.rev.bc --spirv-ext=+SPV_KHR_subgroup_rotate -spirv-text -o - | FileCheck %s --check-prefix=CHECK-SPIRV
+
+// CHECK-SPIRV-DAG: {{[0-9]*}} Capability GroupNonUniformRotateKHR
+// CHECK-SPIRV-DAG: Extension "SPV_KHR_subgroup_rotate"
+
+// CHECK-SPIRV-DAG: TypeInt   [[char:[0-9]+]]   8  0
+// CHECK-SPIRV-DAG: TypeInt   [[short:[0-9]+]]  16 0
+// CHECK-SPIRV-DAG: TypeInt   [[int:[0-9]+]]    32 0
+// CHECK-SPIRV-DAG: TypeInt   [[long:[0-9]+]]   64 0
+// CHECK-SPIRV-DAG: TypeFloat [[half:[0-9]+]]   16
+// CHECK-SPIRV-DAG: TypeFloat [[float:[0-9]+]]  32
+// CHECK-SPIRV-DAG: TypeFloat [[double:[0-9]+]] 64
+
+// CHECK-SPIRV-DAG: Constant [[int]]    [[ScopeSubgroup:[0-9]+]] 3
+// CHECK-SPIRV-DAG: Constant [[char]]   [[char_0:[0-9]+]]        0
+// CHECK-SPIRV-DAG: Constant [[short]]  [[short_0:[0-9]+]]       0
+// CHECK-SPIRV-DAG: Constant [[int]]    [[int_0:[0-9]+]]         0
+// CHECK-SPIRV-DAG: Constant [[int]]    [[int_2:[0-9]+]]         2
+// CHECK-SPIRV-DAG: Constant [[int]]    [[int_4:[0-9]+]]         4
+// CHECK-SPIRV-DAG: Constant [[long]]   [[long_0:[0-9]+]]        0
+// CHECK-SPIRV-DAG: Constant [[half]]   [[half_0:[0-9]+]]        0
+// CHECK-SPIRV-DAG: Constant [[float]]  [[float_0:[0-9]+]]       0
+// CHECK-SPIRV-DAG: Constant [[double]] [[double_0:[0-9]+]]      0
+
+#pragma OPENCL EXTENSION cl_khr_fp16 : enable
+#pragma OPENCL EXTENSION cl_khr_fp64 : enable
+
+// Prototypes from clang's opencl-c.h, as these are not on clang's release/14x branch.
+
+#define __ovld __attribute__((overloadable))
+#define __conv __attribute__((convergent))
+
+char __ovld __conv sub_group_rotate(char, int);
+uchar __ovld __conv sub_group_rotate(uchar, int);
+short __ovld __conv sub_group_rotate(short, int);
+ushort __ovld __conv sub_group_rotate(ushort, int);
+int __ovld __conv sub_group_rotate(int, int);
+uint __ovld __conv sub_group_rotate(uint, int);
+long __ovld __conv sub_group_rotate(long, int);
+ulong __ovld __conv sub_group_rotate(ulong, int);
+float __ovld __conv sub_group_rotate(float, int);
+double __ovld __conv sub_group_rotate(double, int);
+half __ovld __conv sub_group_rotate(half, int);
+
+char __ovld __conv sub_group_clustered_rotate(char, int, uint);
+uchar __ovld __conv sub_group_clustered_rotate(uchar, int, uint);
+short __ovld __conv sub_group_clustered_rotate(short, int, uint);
+ushort __ovld __conv sub_group_clustered_rotate(ushort, int, uint);
+int __ovld __conv sub_group_clustered_rotate(int, int, uint);
+uint __ovld __conv sub_group_clustered_rotate(uint, int, uint);
+long __ovld __conv sub_group_clustered_rotate(long, int, uint);
+ulong __ovld __conv sub_group_clustered_rotate(ulong, int, uint);
+float __ovld __conv sub_group_clustered_rotate(float, int, uint);
+double __ovld __conv sub_group_clustered_rotate(double, int, uint);
+half __ovld __conv sub_group_clustered_rotate(half, int, uint);
+
+// CHECK-SPIRV-LABEL: 5 Function
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[char]] {{[0-9]+}} [[ScopeSubgroup]] [[char_0]] [[int_2]]
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[char]] {{[0-9]+}} [[ScopeSubgroup]] [[char_0]] [[int_2]] [[int_4]]
+// CHECK-SPIRV: FunctionEnd
+
+// CHECK-COMMON-LABEL: @testRotateChar
+
+// CHECK-LLVM: call spir_func i8 @_Z16sub_group_rotateci(i8 0, i32 2)
+// CHECK-LLVM: call spir_func i8 @_Z26sub_group_clustered_rotatecij(i8 0, i32 2, i32 4)
+
+// CHECK-SPV-IR: call spir_func i8 @_Z32__spirv_GroupNonUniformRotateKHRici(i32 3, i8 0, i32 2)
+// CHECK-SPV-IR: call spir_func i8 @_Z32__spirv_GroupNonUniformRotateKHRicij(i32 3, i8 0, i32 2, i32 4)
+kernel void testRotateChar(global char* dst)
+{
+    char v = 0;
+    dst[0] = sub_group_rotate(v, 2);
+    dst[1] = sub_group_clustered_rotate(v, 2, 4);
+}
+
+// CHECK-SPIRV-LABEL: 5 Function
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[char]] {{[0-9]+}} [[ScopeSubgroup]] [[char_0]] [[int_2]]
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[char]] {{[0-9]+}} [[ScopeSubgroup]] [[char_0]] [[int_2]] [[int_4]]
+// CHECK-SPIRV: FunctionEnd
+
+// CHECK-COMMON-LABEL: @testRotateUChar
+
+// CHECK-LLVM: call spir_func i8 @_Z16sub_group_rotateci(i8 0, i32 2)
+// CHECK-LLVM: call spir_func i8 @_Z26sub_group_clustered_rotatecij(i8 0, i32 2, i32 4)
+
+// CHECK-SPV-IR: call spir_func i8 @_Z32__spirv_GroupNonUniformRotateKHRici(i32 3, i8 0, i32 2)
+// CHECK-SPV-IR: call spir_func i8 @_Z32__spirv_GroupNonUniformRotateKHRicij(i32 3, i8 0, i32 2, i32 4)
+kernel void testRotateUChar(global uchar* dst)
+{
+    uchar v = 0;
+    dst[0] = sub_group_rotate(v, 2);
+    dst[1] = sub_group_clustered_rotate(v, 2, 4);
+}
+
+// CHECK-SPIRV-LABEL: 5 Function
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[short]] {{[0-9]+}} [[ScopeSubgroup]] [[short_0]] [[int_2]]
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[short]] {{[0-9]+}} [[ScopeSubgroup]] [[short_0]] [[int_2]] [[int_4]]
+// CHECK-SPIRV: FunctionEnd
+
+// CHECK-COMMON-LABEL: @testRotateShort
+
+// CHECK-LLVM: call spir_func i16 @_Z16sub_group_rotatesi(i16 0, i32 2)
+// CHECK-LLVM: call spir_func i16 @_Z26sub_group_clustered_rotatesij(i16 0, i32 2, i32 4)
+
+// CHECK-SPV-IR: call spir_func i16 @_Z32__spirv_GroupNonUniformRotateKHRisi(i32 3, i16 0, i32 2)
+// CHECK-SPV-IR: call spir_func i16 @_Z32__spirv_GroupNonUniformRotateKHRisij(i32 3, i16 0, i32 2, i32 4)
+kernel void testRotateShort(global short* dst)
+{
+    short v = 0;
+    dst[0] = sub_group_rotate(v, 2);
+    dst[1] = sub_group_clustered_rotate(v, 2, 4);
+}
+
+// CHECK-SPIRV-LABEL: 5 Function
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[short]] {{[0-9]+}} [[ScopeSubgroup]] [[short_0]] [[int_2]]
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[short]] {{[0-9]+}} [[ScopeSubgroup]] [[short_0]] [[int_2]] [[int_4]]
+// CHECK-SPIRV: FunctionEnd
+
+// CHECK-COMMON-LABEL: @testRotateUShort
+
+// CHECK-LLVM: call spir_func i16 @_Z16sub_group_rotatesi(i16 0, i32 2)
+// CHECK-LLVM: call spir_func i16 @_Z26sub_group_clustered_rotatesij(i16 0, i32 2, i32 4)
+
+// CHECK-SPV-IR: call spir_func i16 @_Z32__spirv_GroupNonUniformRotateKHRisi(i32 3, i16 0, i32 2)
+// CHECK-SPV-IR: call spir_func i16 @_Z32__spirv_GroupNonUniformRotateKHRisij(i32 3, i16 0, i32 2, i32 4)
+kernel void testRotateUShort(global ushort* dst)
+{
+    ushort v = 0;
+    dst[0] = sub_group_rotate(v, 2);
+    dst[1] = sub_group_clustered_rotate(v, 2, 4);
+}
+
+// CHECK-SPIRV-LABEL: 5 Function
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[int]] {{[0-9]+}} [[ScopeSubgroup]] [[int_0]] [[int_2]]
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[int]] {{[0-9]+}} [[ScopeSubgroup]] [[int_0]] [[int_2]] [[int_4]]
+// CHECK-SPIRV: FunctionEnd
+
+// CHECK-COMMON-LABEL: @testRotateInt
+
+// CHECK-LLVM: call spir_func i32 @_Z16sub_group_rotateii(i32 0, i32 2)
+// CHECK-LLVM: call spir_func i32 @_Z26sub_group_clustered_rotateiij(i32 0, i32 2, i32 4)
+
+// CHECK-SPV-IR: call spir_func i32 @_Z32__spirv_GroupNonUniformRotateKHRiii(i32 3, i32 0, i32 2)
+// CHECK-SPV-IR: call spir_func i32 @_Z32__spirv_GroupNonUniformRotateKHRiiij(i32 3, i32 0, i32 2, i32 4)
+kernel void testRotateInt(global int* dst)
+{
+    int v = 0;
+    dst[0] = sub_group_rotate(v, 2);
+    dst[1] = sub_group_clustered_rotate(v, 2, 4);
+}
+
+// CHECK-SPIRV-LABEL: 5 Function
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[int]] {{[0-9]+}} [[ScopeSubgroup]] [[int_0]] [[int_2]]
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[int]] {{[0-9]+}} [[ScopeSubgroup]] [[int_0]] [[int_2]] [[int_4]]
+// CHECK-SPIRV: FunctionEnd
+
+// CHECK-COMMON-LABEL: @testRotateUInt
+
+// CHECK-LLVM: call spir_func i32 @_Z16sub_group_rotateii(i32 0, i32 2)
+// CHECK-LLVM: call spir_func i32 @_Z26sub_group_clustered_rotateiij(i32 0, i32 2, i32 4)
+
+// CHECK-SPV-IR: call spir_func i32 @_Z32__spirv_GroupNonUniformRotateKHRiii(i32 3, i32 0, i32 2)
+// CHECK-SPV-IR: call spir_func i32 @_Z32__spirv_GroupNonUniformRotateKHRiiij(i32 3, i32 0, i32 2, i32 4)
+kernel void testRotateUInt(global uint* dst)
+{
+    uint v = 0;
+    dst[0] = sub_group_rotate(v, 2);
+    dst[1] = sub_group_clustered_rotate(v, 2, 4);
+}
+
+// CHECK-SPIRV-LABEL: 5 Function
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[long]] {{[0-9]+}} [[ScopeSubgroup]] [[long_0]] [[int_2]]
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[long]] {{[0-9]+}} [[ScopeSubgroup]] [[long_0]] [[int_2]] [[int_4]]
+// CHECK-SPIRV: FunctionEnd
+
+// CHECK-COMMON-LABEL: @testRotateLong
+
+// CHECK-LLVM: call spir_func i64 @_Z16sub_group_rotateli(i64 0, i32 2)
+// CHECK-LLVM: call spir_func i64 @_Z26sub_group_clustered_rotatelij(i64 0, i32 2, i32 4)
+
+// CHECK-SPV-IR: call spir_func i64 @_Z32__spirv_GroupNonUniformRotateKHRili(i32 3, i64 0, i32 2)
+// CHECK-SPV-IR: call spir_func i64 @_Z32__spirv_GroupNonUniformRotateKHRilij(i32 3, i64 0, i32 2, i32 4)
+kernel void testRotateLong(global long* dst)
+{
+    long v = 0;
+    dst[0] = sub_group_rotate(v, 2);
+    dst[1] = sub_group_clustered_rotate(v, 2, 4);
+}
+
+// CHECK-SPIRV-LABEL: 5 Function
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[long]] {{[0-9]+}} [[ScopeSubgroup]] [[long_0]] [[int_2]]
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[long]] {{[0-9]+}} [[ScopeSubgroup]] [[long_0]] [[int_2]] [[int_4]]
+// CHECK-SPIRV: FunctionEnd
+
+// CHECK-COMMON-LABEL: @testRotateULong
+
+// CHECK-LLVM: call spir_func i64 @_Z16sub_group_rotateli(i64 0, i32 2)
+// CHECK-LLVM: call spir_func i64 @_Z26sub_group_clustered_rotatelij(i64 0, i32 2, i32 4)
+
+// CHECK-SPV-IR: call spir_func i64 @_Z32__spirv_GroupNonUniformRotateKHRili(i32 3, i64 0, i32 2)
+// CHECK-SPV-IR: call spir_func i64 @_Z32__spirv_GroupNonUniformRotateKHRilij(i32 3, i64 0, i32 2, i32 4)
+kernel void testRotateULong(global ulong* dst)
+{
+    ulong v = 0;
+    dst[0] = sub_group_rotate(v, 2);
+    dst[1] = sub_group_clustered_rotate(v, 2, 4);
+}
+
+// CHECK-SPIRV-LABEL: 5 Function
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[float]] {{[0-9]+}} [[ScopeSubgroup]] [[float_0]] [[int_2]]
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[float]] {{[0-9]+}} [[ScopeSubgroup]] [[float_0]] [[int_2]] [[int_4]]
+// CHECK-SPIRV: FunctionEnd
+
+// CHECK-COMMON-LABEL: @testRotateFloat
+
+// CHECK-LLVM: call spir_func float @_Z16sub_group_rotatefi(float 0.000000e+00, i32 2)
+// CHECK-LLVM: call spir_func float @_Z26sub_group_clustered_rotatefij(float 0.000000e+00, i32 2, i32 4)
+
+// CHECK-SPV-IR: call spir_func float @_Z32__spirv_GroupNonUniformRotateKHRifi(i32 3, float 0.000000e+00, i32 2)
+// CHECK-SPV-IR: call spir_func float @_Z32__spirv_GroupNonUniformRotateKHRifij(i32 3, float 0.000000e+00, i32 2, i32 4)
+kernel void testRotateFloat(global float* dst)
+{
+    float v = 0;
+    dst[0] = sub_group_rotate(v, 2);
+    dst[1] = sub_group_clustered_rotate(v, 2, 4);
+}
+
+// CHECK-SPIRV-LABEL: 5 Function
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[half]] {{[0-9]+}} [[ScopeSubgroup]] [[half_0]] [[int_2]]
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[half]] {{[0-9]+}} [[ScopeSubgroup]] [[half_0]] [[int_2]] [[int_4]]
+// CHECK-SPIRV: FunctionEnd
+
+// CHECK-COMMON-LABEL: @testRotateHalf
+
+// CHECK-LLVM: call spir_func half @_Z16sub_group_rotateDhi(half 0xH0000, i32 2)
+// CHECK-LLVM: call spir_func half @_Z26sub_group_clustered_rotateDhij(half 0xH0000, i32 2, i32 4)
+
+// CHECK-SPV-IR: call spir_func half @_Z32__spirv_GroupNonUniformRotateKHRiDhi(i32 3, half 0xH0000, i32 2)
+// CHECK-SPV-IR: call spir_func half @_Z32__spirv_GroupNonUniformRotateKHRiDhij(i32 3, half 0xH0000, i32 2, i32 4)
+kernel void testRotateHalf(global half* dst)
+{
+    half v = 0;
+    dst[0] = sub_group_rotate(v, 2);
+    dst[1] = sub_group_clustered_rotate(v, 2, 4);
+}
+
+// CHECK-SPIRV-LABEL: 5 Function
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[double]] {{[0-9]+}} [[ScopeSubgroup]] [[double_0]] [[int_2]]
+// CHECK-SPIRV: GroupNonUniformRotateKHR [[double]] {{[0-9]+}} [[ScopeSubgroup]] [[double_0]] [[int_2]] [[int_4]]
+// CHECK-SPIRV: FunctionEnd
+
+// CHECK-COMMON-LABEL: @testRotateDouble
+
+// CHECK-LLVM: call spir_func double @_Z16sub_group_rotatedi(double 0.000000e+00, i32 2)
+// CHECK-LLVM: call spir_func double @_Z26sub_group_clustered_rotatedij(double 0.000000e+00, i32 2, i32 4)
+
+// CHECK-SPV-IR: call spir_func double @_Z32__spirv_GroupNonUniformRotateKHRidi(i32 3, double 0.000000e+00, i32 2)
+// CHECK-SPV-IR: call spir_func double @_Z32__spirv_GroupNonUniformRotateKHRidij(i32 3, double 0.000000e+00, i32 2, i32 4)
+kernel void testRotateDouble(global double* dst)
+{
+    double v = 0;
+    dst[0] = sub_group_rotate(v, 2);
+    dst[1] = sub_group_clustered_rotate(v, 2, 4);
+}
-- 
2.20.1

